client/sql/dataframewriter.go

```diff
@@ -13,15 +16,27 @@ type DataFrameWriter interface {
 	// Format specifies data format (data source type) for the underlying data, e.g. parquet.
 	Format(source string) DataFrameWriter
 	// Save writes data frame to the given path.
-	Save(path string) error
+	Save(ctx context.Context, path string) error
+}
+
+type sparkExecutor interface {
+	executePlan(ctx context.Context, plan *proto.Plan) (*executePlanClient, error)
+	analyzePlan(ctx context.Context, plan *proto.Plan) (*proto.AnalyzePlanResponse, error)
+}
```

Comment:
the code location of this type is kind of weird. Again not a go expert here.

In Python we use the following abstraction:

DataFrame -> Session -> Client

Can we keep the naming somewhat consistent and does this need to be in this file?

Reply:
Absolutely, but I think we will need to figure out exactly what the right placement is.

By introducing a private interface type, I can declare that this code depends on something that has this signature instead of having a circular dependency directly back on the implementation code for the session.

Design-wise, I would probably prefer to separate the business logic code that exposes the Sql function from the executor code that uses the client and consider those as two different layers. It is somewhat messy that the session creates a dataframe that then in turn depends on the session. Go does not prevent this type of circular dependency when it happens within the same package but this would be a compilation error if we put dataframe in its own package.

We could actually consider doing that, specifically to ensure that the design becomes/remains layered in the way that you describe. Wdyt?

Reply:
If you want (and it makes better use of your time than comments here), we can set up a sync session on zoom/meet sometime this week. We are in the same tz